### PR TITLE
Implement intrinsic `index`

### PR DESCRIPTION
**Pull request by HarshitaKalani**

### CODE REVIEW

**Review by certik — APPROVED** [review]

**Comment by Pranavchiku** [comment]
The CI error is unrelated, rebase this, resolve conflicts and this will be good to merge.

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -4053,6 +4054,134 @@ namespace StringContainsSet {
...
+            b.Assignment(found, args[2])
+        }, {
+            b.Assignment(found, b.bool32(1))
+        }));
+        while_loop_body.push_back(b.Assignment(i, b.i_tAdd(i, b.i(1, return_type), return_type)));
+                
+        body.push_back(al, b.While(b.And(b.iLt(i, str_len), b.boolEq(found, b.bool32(1))), while_loop_body));
+
+        ASR::symbol_t *f_sym = make_ASR_Function_t(fn_name, fn_symtab, dep, args,
+            body, idx, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:4178]: I'll prefer to write this as:


```c
body.push_back(al, b.While(b.And(b.iLt(i, str_len), b.boolEq(found, b.bool32(1))), {
// write down all while loop body statements here, those will be `,` separated, this makes it more readable
}));
```

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -4053,6 +4054,134 @@ namespace StringContainsSet {
...
+                end do
+            end function
+        */
+        body.push_back(al, b.Assignment(idx, b.i(0, return_type)));
+        body.push_back(al, b.Assignment(i, b.i(1, return_type)));
+        body.push_back(al, b.Assignment(found, b.bool32(1)));
+        ASR::expr_t* str_len = b.StringLen(args[0]);
+        ASR::expr_t* sub_len = b.StringLen(args[1]);
+        ASR::expr_t* string_section = b.StringSection(args[0], b.i_tSub(pos, b.i(1, return_type), return_type), pos);
+        ASR::expr_t* substring_section = b.StringSection(args[1], b.i_tSub(j, b.i(1, return_type), return_type), j);
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:4149]: If we do not need to use these more than once, it needs not to be declared as a separate variable

File: src/libasr/pass/intrinsic_functions.h (hunk 1 above)
Comment (HarshitaKalani) [reply src/libasr/pass/intrinsic_functions.h]: I think this looks better. Will update the same for `verify` and `scan` in a separate PR if the idea looks good to you as well.

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -4179,6 +4180,129 @@ namespace StringFindSet {
...
+                b.Assignment(k, b.i_tAdd(k, b.i(1, return_type), return_type)),
+            }),
+            b.If(b.boolEq(found, b.bool32(1)), {
+                b.Assignment(idx, i),
+                b.Assignment(found, args[2])
+            }, {
+                b.Assignment(found, b.bool32(1))
+            }),
+            b.Assignment(i, b.i_tAdd(i, b.i(1, return_type), return_type)),
+        }));
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:4296]: This is good, thanks for doing this. This way one can easily understand.

**Review by Pranavchiku — COMMENTED** [review]
Thanks for this PR. You may wish to look at #3671 to have an idea about implementing complex intrinsic, still not sure if that is right approach.

**Comment by HarshitaKalani** [comment]
Thanks for this

**Pull request merged**